code/controllers/autofire.dm

```diff
@@ -0,0 +1,264 @@
+		SSautomatedfire.bucket_count--
+		var/bucket_pos = BUCKET_POS(next_fire)
+		if (bucket_pos > 0)
+			var/datum/component/automatedfire/bucket_head = bucket_list[bucket_pos]
+			if (bucket_head == src)
+				bucket_list[bucket_pos] = next
+		if (prev != next)
+			prev.next = next
+			next.prev = prev
+		else
+			prev?.next = null
+			next?.prev = null
+		prev = next = null
+		next_fire = new_next_fire
+
+	// Ensure the next_fire time is properly bound to avoid missing a scheduled event
+	next_fire = max(CEILING(next_fire, world.tick_lag), world.time + world.tick_lag)
+
+	// Get bucket position and a local reference to the datum var, it's faster to access this way
+	var/bucket_pos = BUCKET_POS(next_fire)
+
+	// Get the bucket head for that bucket, increment the bucket count
+	var/datum/component/automatedfire/bucket_head = bucket_list[bucket_pos]
+	SSautomatedfire.bucket_count++
+
+	// If there is no existing head of this bucket, we can set this shooter to be that head
+	if (!bucket_head)
+		bucket_list[bucket_pos] = src
+		return
+
+	// Otherwise it's a simple insertion into the circularly doubly-linked list
+	if (!bucket_head.prev)
+		bucket_head.prev = bucket_head
+	next = bucket_head
+	prev = bucket_head.prev
+	next.prev = src
+	prev.next = src
+
+
+/**
+ * Removes this autofire component from the autofire subsystem
+ */
+/datum/component/automatedfire/proc/unschedule_shot() //This is probably not needed
+	// Attempt to find the bucket that contains this component
+	var/bucket_pos = BUCKET_POS(next_fire)
+
+	// Get local references to the subsystem's vars, faster than accessing on the datum
+	var/list/bucket_list = SSautomatedfire.bucket_list
+
+	// Attempt to get the head of the bucket
+	var/datum/component/automatedfire/bucket_head
+	if (bucket_pos > 0)
+		bucket_head = bucket_list[bucket_pos]
+
+	//Replace the bucket head if needed
+	if(bucket_head == src)
+		bucket_list[bucket_pos] = next
+	SSautomatedfire.bucket_count--
+	// Remove the shooter from the bucket, ensuring to maintain
+	// the integrity of the bucket's list if relevant
+	if(prev != next)
+		prev.next = next
+		next.prev = prev
+	else
+		prev?.next = null
+		next?.prev = null
+	prev = next = null
+
+///Handle the firing of the autofire component
+/datum/component/automatedfire/proc/process_shot()
+	return
+
+#undef BUCKET_LEN
+#undef BUCKET_POS
+
+/obj/structure/turret_debug
+	name = "debug turret"
+	///What kind of ammo it uses
+	var/datum/ammo/ammo
+	///Its target
+	var/atom/target 
+	///At wich rate it fires in ticks
+	var/firerate = 5
+
+/obj/structure/turret_debug/fast
+	name = "debug turret fast"
+	firerate = 1
+
+/obj/structure/turret_debug/slow
+	name = "debug turret slow"
+	firerate = 25
+
+/obj/structure/turret_debug/Initialize()
+	. = ..()
+	ammo = GLOB.ammo_list[/datum/ammo/xeno/acid]
+	target = locate(x+5, y, z)
+	AddComponent(/datum/component/automatedfire/automatic_shoot_at, firerate, ammo)
+	SEND_SIGNAL(src, COMSIG_AUTOMATIC_SHOOTER_START_SHOOTING_AT, target)
```

Comment:
Signals are not supposed to just replace proc calls. They're generic events.
Since this s only used by `/automatedfire` to start shooting it should probably be an initialization parameter instead.

Reply:
Yes of course, but this was just a debug turret. I used the same structure than the xenos turret, and these needs to be able to stop and start shooting. I do believe signals are the way to go to communicate between components and parents?

Reply:
As bravemole said this is for the debug turret